### ISSUE TITLE
CPU Basecalling Extremely Slow

### ISSUE THREAD

**Issue by scottgigante** [nanoporetech/nanonet#47]
Hi,

I've trained a neural network with a 5-base alphabet, and have tried this with both 5mers and 3mers. The training runs well, but basecalling (CPU only) is extraordinarily slow. For example, I have 100,000 reads I am trying to basecall, and after 24 days of CPU time (running on 40 threads, but only about 14 are active) I've basecalled approximately a tenth of the files.

Other arguments: --chemistry r9, --section template. Everything else default.

Cheers,
Scott

**Comment by cjw85** [comment]
Hi @scottgigante,

Do the completed basecalls look sane? If not we have have inadvertently assumed 4 bases at some point in the code. That should be considered a bug because we had intended it to be generic.

Have you tried profiling the code and comparing with standard models (with e.g. kernprof)? I would not have expected at either basis set to lead to egregiously slow basecalling. The RNN step should not be adversely affected, it is only the final classification layer computation that is engrossed. The decoding step will scale with the size of the basis set, so if anything your second case should be faster that the default models. Four bases is however convenient for SSE vectorization in the code, but that is likely a minor factor.

BTW I'm not sure how the --chemistry and --model options interact to produce a resultant set of argument values. You might be best served by placing an additional entry here: https://github.com/nanoporetech/nanonet/blob/master/nanonet/nanonetcall.py#L36. (The --chemistry option was added as a convenience for specifying self-consistent sets of arguments).

**Comment by scottgigante** [comment]
Hi @cjw85,

I've run kernprof (and in doing so, had to make nanonetcall run as a python script rather than as an executable) - running in this way, the 5-base 3mer indeed runs faster than than 4-base 5mer and has blockages around the same places. Not sure if posting the output will be useful - it is exactly as you would expect.

Kernprof doesn't play nice with multiprocessing, so I could only do this on a single thread. However, I did notice this when running without kernprof (still as a script rather than an application):

```
% nanonet/nanonetcall.py --chemistry r9 --jobs 1 --model ../nanomod/data/lucattini_no_phage_3mer_best.npy --output prof_test/5base3mer.fasta --section template prof_test/reads
Basecalled 50 reads (198362 bases, 505841 events) in 127.160819054s (wall time)
Run network:  79.49 ( 2.495 kb/s,  6.364 kev/s)
Decoding:     35.98 ( 5.513 kb/s, 14.058 kev/s)

% nanonet/nanonetcall.py --chemistry r9 --jobs 1 --model ../nanomod/models/r9_template.npy --output prof_test/4base5mer.fasta --section template prof_test/reads
Basecalled 50 reads (284874 bases, 505841 events) in 216.724620104s (wall time)
Run network:  84.61 ( 3.367 kb/s,  5.978 kev/s)
Decoding:     96.72 ( 2.945 kb/s,  5.230 kev/s)

nanonet/nanonetcall.py --chemistry r9 --jobs 4 --model ../nanomod/data/lucattini_no_phage_3mer_best.npy --output prof_test/5base3mer.fasta --section template prof_test/reads
Basecalled 50 reads (198362 bases, 505841 events) in 123.744063139s (wall time)
Run network: 369.17 ( 0.537 kb/s,  1.370 kev/s)
Decoding:     88.17 ( 2.250 kb/s,  5.737 kev/s)

% nanonet/nanonetcall.py --chemistry r9 --jobs 4 --model ../nanomod/models/r9_template.npy --output prof_test/4base5mer.fasta --section template prof_test/reads
Basecalled 50 reads (284874 bases, 505841 events) in 144.387593031s (wall time)
Run network: 333.73 ( 0.854 kb/s,  1.516 kev/s)
Decoding:    156.01 ( 1.826 kb/s,  3.242 kev/s)

% nanonet/nanonetcall.py --chemistry r9 --jobs 16 --model ../nanomod/data/lucattini_no_phage_3mer_best.npy --output prof_test/5base3mer.fasta --section template prof_test/reads
Basecalled 50 reads (198362 bases, 505841 events) in 117.415096045s (wall time)
Run network: 1347.74 ( 0.147 kb/s,  0.375 kev/s)
Decoding:    282.38 ( 0.702 kb/s,  1.791 kev/s)

% nanonet/nanonetcall.py --chemistry r9 --jobs 16 --model ../nanomod/models/r9_template.npy --output prof_test/4base5mer.fasta --section template prof_test/reads
Basecalled 50 reads (284874 bases, 505841 events) in 156.632887125s (wall time)
Run network: 1433.27 ( 0.199 kb/s,  0.353 kev/s)
Decoding:    612.48 ( 0.465 kb/s,  0.826 kev/s)
```

It's true that it's not very fair to test the threading performance on such a small test set, but the point stands - performance drops rapidly as ```--jobs``` increases. 

As a side note, ```--chemistry``` and ```--model``` play nice together - the model overwrites the default given by chemistry.

**Comment by scottgigante** [comment]
Further to this concern, I have run a real dataset with my own models and the original ```r9_template.npy``` model provided by nanonet. Each job was launched roughly 24 hours ago on a VM with 40 cores. The jobs are currently incomplete, and while fasta size is not an entirely accurate measure of basecalling speed *between* models due to differing error rates, it is quite indicative within models.
```
-rw-r--r-- 1 gigante.s allstaff  81669955 Apr  5 14:32 data/nanonet.5base.5mer.16.fasta
-rw-r--r-- 1 gigante.s allstaff  84608064 Apr  5 14:32 data/nanonet.5base.5mer.40.fasta
-rw-r--r-- 1 gigante.s allstaff  86581686 Apr  5 14:32 data/nanonet.5base.5mer.8.fasta
-rw-r--r-- 1 gigante.s allstaff 106647106 Apr  5 14:32 data/nanonet.5base.3mer.16.fasta
-rw-r--r-- 1 gigante.s allstaff 106920053 Apr  5 14:32 data/nanonet.5base.3mer.40.fasta
-rw-r--r-- 1 gigante.s allstaff 107460325 Apr  5 14:32 data/nanonet.5base.3mer.8.fasta
-rw-r--r-- 1 gigante.s allstaff 130054568 Apr  5 14:32 data/nanonet.4base.5mer.40.fasta
-rw-r--r-- 1 gigante.s allstaff 134330328 Apr  5 14:32 data/nanonet.4base.5mer.16.fasta
-rw-r--r-- 1 gigante.s allstaff 136592433 Apr  5 14:32 data/nanonet.4base.5mer.8.fasta
-rwxrwxr-x 1 gigante.s allstaff 706433481 Mar 24 19:04 data/metrichor.fasta
```
It is clear that on a system with 40 cores, running with just 8 "jobs" is preferable to running with more. Note that in ```top```, anecdotally all 40 cores are in constant use, have been for 24 hours, and in the best run we are only about 20% of the way through the jobs (based on the expected size of the fasta from metrichor). 5 days at 40 cores is a long time to basecall some 120000 reads.

**Comment by cjw85** [comment]
In all the benchmarking we have done, the performance scales well with the number of physical cores. On machines with simultaneous multithreading (such as Intel's Hyperthreading), it is best to restrict `--jobs` to the number of physical cores.

Are you using a multithreaded BLAS? This generally has a detrimental performance. You should set the BLAS threads to 1; depending which implementation you are using this can be controlled with an environment variable such as: `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, or `OPENBLAS_NUM_THREADS`.

We have some benchmarking results here:
https://github.com/nanoporetech/nanonet#benchmarks

Though nanonet's numpy implementation of the BLSTM is no couch potato, it will be slower than the more direct C++ implementation in Albacore. Unfortunately I believe Albacore will not be able to run basecalling with your models.

**Comment by scottgigante** [comment]
@cjw85 finally managed to get OpenBLAS running and try this on single-threaded BLAS - we were running ATLAS where the number of threads is specified at compile time. This indeed solves the problem. Thanks!
